src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -796,14 +798,36 @@ pub enum StaleItem {
 impl LocalFingerprint {
     /// Read the environment variable of the given env `key`, and creates a new
     /// [`LocalFingerprint::RerunIfEnvChanged`] for it.
-    ///
-    // TODO: This is allowed at this moment. Should figure out if it makes
-    // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        env_config: &Arc<HashMap<String, OsString>>,
+        env_config_insensitive: &Arc<LazyCell<HashMap<String, OsString>>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+        let val = if let Some(val) = env::var(key).ok() {
+            Some(val)
+        } else {
+            match env_config.get(key) {
+                Some(value) => value.to_str().map(|s| s.to_string()),
+                None => {
+                    if cfg!(windows) {
+                        env_config_insensitive
```

Comment:
Rather than having to do all of this extra work, I wonder if we should make the env map itself case insensitive.  Looking at how `env_config()` is used, I think we have several places where we don't handle case sensitivity and maybe we should?

What we could do is define a type def for `EnvMap` where the key is `unicase::Ascii<String>` and have the global context return that.

@weihanglo thoughts?

Reply:
That should be conditioned on platform, maybe only on Windows?

Cargo does that upfront for environment variables from configuration enviornment: https://github.com/rust-lang/cargo/blob/master/src/cargo/util/context/environment.rs

Reply:
Sorry, Yes, I meant conditioned on the platform (which is why the typedef would be there)

Reply:
Hmm, if we also have a normalization process, I wonder if we should apply that more generally